core/src/test/scala/org/apache/spark/sql/store/CreateIndexTest.scala

```diff
@@ -0,0 +1,772 @@
+        s"where $table1.col1 = $table3.col1") {
+      validateIndex(Seq(index1), table3)(_)
+    }
+
+    executeQ(s"select * from $table1 t1 /*+ ${QueryHint.Index}($index1) */, $table3 t2 " +
+        s"where t1.col1 = t2.col1") {
+      validateIndex(Seq(index1), table3)(_)
+    }
+
+    executeQ(s"select * from $table1 /*+ ${QueryHint.Index}($index1) */ as t1, $table3 t2 " +
+        s"where t1.col1 = t2.col1") {
+      validateIndex(Seq(index1), table3)(_)
+    }
+
+    executeQ(s"select * from $table1 tab1 join $table2 tab2 on tab1.col2 = tab2.col2") {
+      validateIndex(Seq.empty, table1, table2)(_)
+    }
+
+    try {
+      executeQ(s"select * from (select * from $table1 ) t1 /*+ ${QueryHint.Index}($index1) */" +
+          s", $table3 t2 where t1.col1 = t2.col1")
+      fail(s"exepected AnalysisException as ${QueryHint.Index}" +
+          s" hint cannot be specified on derived tables")
+    } catch {
+      case ae: AnalysisException if ae.message.contains(s"${QueryHint.Index}") =>
+      // okay correct message.
+      case e: Throwable => alert(e.getMessage)
+        throw e
+    }
+  }
+
+  test("Test two table joins corner cases") {
+
+    val table1 = "tabOne"
+    val table2 = "tabTwo"
+    val table3 = "tabThree"
+
+    val index1 = s"${table1}_IdxOne"
+    val index2 = s"${table1}_IdxTwo"
+
+    val index31 = s"${table3}_IdxOne"
+
+    val snContext = SnappyContext(sc)
+    context.set(snContext)
+    snContext.setConf(SQLConf.AUTO_BROADCASTJOIN_THRESHOLD.key, "-1")
+
+    createBase3Tables(snContext, table1, table2, table3)
+
+    doPrint("Verify index create and drop for various index types")
+    snContext.sql(s"create index $index1 on $table1 " +
+        s" (COL1) Options (colocate_with  '$table3')")
+    indexesToDrop += index1
+
+    snContext.sql(s"create index $index2 on $table1 " +
+        s" (COL2, COL3) Options (colocate_with  '$table2')")
+    indexesToDrop += index2
+
+    snContext.sql(s"create index $index31 on $table3 " +
+        s" (COL2, COL3) Options (colocate_with  '$index2')")
+    indexesToDrop += index31
+
+    val executeQ = QueryExecutor(snContext)
+
+    executeQ(s"select * from $table1 tab1 join $table3 tab2 on tab1.col1 = tab2.col1 " +
+        s"where tab1.col1 = 111 ") {
+      validateIndex(Seq(index1), table3)(_)
+    }
+
+    executeQ(s"select $table1.col2, $table3.col3 from $table1, $table3 " +
+        s"where $table1.col1 = $table3.col1 and $table3.col1 = 111 ") {
+      validateIndex(Seq(index1), table3)(_)
+    }
+
+    executeQ(s"select t1.col2, t2.col3 from $table1 t1 join $table2 t2 on t1.col2 = t2.col2 " +
+        s"where t1.col3 = t2.col3 ", true) {
+      validateIndex(Seq(index2), table2)(_)
+    }
+
+    executeQ(s"select t1.col2, t2.col3 from $table1 t1 join $table2 t2 on t1.col2 = t2.col2 " +
+        s"where t1.col2 = t2.col2 and t1.col3 = t2.col3 ") { df =>
+      // validateIndex(Seq(index2", table2"))(_)
```

Comment:
@ymahajan good catch. No here join conditions are repeating on col2 but that shouldn't stop picking up the index. Lost in translation.. will enable it and check.